### PR TITLE
DE-135 - Doppler security

### PR DESCRIPTION
**Pull request by andresmoschini**
Hi team, 

This is a big PR! but I think that it is possible to review it easily by reading commit by commit.

It supports 3 different scenarios, not all APIs require all of them:

* Verify token signature (whatever the account or superuser status is)
* Verify if the token is from a superuser
* Verify if the user has access to an URL based on his account ID or name

Maybe you do not need all of these scenarios and having a specific policy as the default is enough.

The tests are very detailed, so, it could be easy and safe adapt it for a new behavior.

---

Tested in INT environment with my account's token:

image

### CODE REVIEW

File: Doppler.HelloMicroservice/DopplerSecurity/DopplerSecurityServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                            AllowOwnResource = true
+                        })
+                        .RequireAuthenticatedUser()
+                        .Build();
+
+                    // TODO: I would like to use ownResourceOrSuperUserPolicy as the default policy, but I
+                    // cannot override a more restrictive policy with a less restrictive one. So,
+                    // for the moment, we have to be carefull and chooses the right one for each
+                    // controller.
+                    o.DefaultPolicy = simpleAuthenticationPolicy;
```
Comment (andresmoschini) [inline Doppler.HelloMicroservice/DopplerSecurity/DopplerSecurityServiceCollectionExtensions.cs:49]: @ms-darianbenito this what we were talking about today.

**Comment by andresmoschini** [comment]
Tested in INT environment with my account's token:

image

File: Doppler.HelloMicroservice/private-keys-dev/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+The steps needed to do so are the following
+
+1. Go to [jwt.io](https://jwt.io/)
+2. Select RS256 as the algorithm
+3. Enter the payload
+4. Paste the key mentioned above in the private key section
+5. [jwt.io](https://jwt.io/) will generate a token ready to be used in the tests
+
+**Note:** The tokens generated for unit tests will be valid for that context only
```
Comment (ms-darianbenito) [inline Doppler.HelloMicroservice/private-keys-dev/README.md:13]: Lo veo bien como parte de este commit. También quedaría claro en un siguiente commint con `docs:`

File: Doppler.HelloMicroservice.Test/AuthorizationTest.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        public async Task GET_helloAnonymous_should_not_require_token(string url, HttpStatusCode expectedStatusCode)
+        {
+            // Arrange
+            var client = _factory.CreateClient(new WebApplicationFactoryClientOptions());
+
+            // Act
+            var response = await client.GetAsync(url);
+
+            // Assert
+            Assert.Equal(expectedStatusCode, response.StatusCode);
```
Comment (ms-darianbenito) [inline Doppler.HelloMicroservice.Test/AuthorizationTest.cs:48]: I am not sure but we need that anonymous endpoint just not response `401` status code.
`Assert.NotEqual(HttpStatusCode.Unauthorized, response.StatusCode);`

Comment (andresmoschini) [reply Doppler.HelloMicroservice.Test/AuthorizationTest.cs:48]: I did not understand.

Is not enough waiting for a `200 OK`?

Comment (ms-darianbenito) [reply Doppler.HelloMicroservice.Test/AuthorizationTest.cs:48]: Yes, It is, but I am thinking on another response status like `InternalServerError`, and that still a valid scenario.

Comment (andresmoschini) [reply Doppler.HelloMicroservice.Test/AuthorizationTest.cs:48]: Oh, I understand now.

I think that an `InternalServerError` is not enough to confirm that anonymous authorization is working fine.

Maybe there is another problem, so, in that case, the result should not be pass, maybe inconclusive, but in that case, the test will be more complex.

**Pull request merged**